[EPIC] v7.6.1 Release

derekpierre opened the pull request:
**Type of PR:**
- [ ] Bugfix
- [ ] Feature
- [ ] Documentation
- [x] Other

**Required reviews:** 
- [ ] 1
- [x] 2
- [ ] 3

**What this does:**
Potential for a v7.6.1 release. If we decide otherwise any code can go into v7.7.x without the bumpversion commit.

PRs:
- [x] #3647 
- [x] #3648
- [x] #3640
- [x] #3649 
- [x] #3650

cygnusv committed:
TACo favicon in status page

cygnusv committed:
Improve logging of handover

cygnusv committed:
Tone down discovery loop logging on soft errors

derekpierre committed:
Bump version: 7.6.0 → 7.6.1

derekpierre committed:
Use debian trixie as the base rust image for our rust-python image.

derekpierre committed:
Use updated rust-python:3.12.11 that is based on debian trixie for `nucypher` docker image builds.

derekpierre committed:
Add feature newsfragment for #3643.

derekpierre committed:
Set max chunk size to 1,000 instead of 10,000.

derekpierre committed:
Handle issues during event scanning better with block range adjustments, retries and delays.
The EventScanner has to be adaptable since rpc endpoints can experience issues retrieving events due to rate/size limit/block range restrictions.

derekpierre committed:
Re-use chunk size that successfully worked after decrease.

derekpierre committed:
Add additional check and adjustment when encountering alchemy free tier event logs limits.
Add a few env variables to provide us with some additional flexibility.

Co-authored-by: Manuel Montenegro <manuel@nucypher.com>
Co-authored-by: KPrasch <kieranprasch@gmail.com>

derekpierre committed:
Add enforcement of min/max scan chunk size for EventScanner.
Updated unit tests accordingly.

derekpierre committed:
Add tests for _get_logs processing including http errors, retries and alchemy free tier checks.

manumonti committed:
Apply ruff formatting

manumonti committed:
Enable json error test

manumonti committed:
Add test for get logs when connection error

derekpierre committed:
Rename `i` variable to `attempt` for clarity.

derekpierre committed:
Add bugfix newsfragment for #3647.

derekpierre committed:
One environment variable to rule them all (for dkg ritual). Allow setting of min chunk size specifically for DKG rituals that basically allows overriding min chunk size for event scanning.

derekpierre committed:
Used named constants for default parameters used by EventScanner.
Update method documentation.

derekpierre committed:
Update event scanning tests to use higher level call to scan_chunk() instead of lower level _get_log() call; this improves test coverage.
Add test to ensure that alchemy free tier is only retried once regardless of number of retries, since retrying doesn't change the chunk size.

derekpierre committed:
Rename parameters and constants for clarity.

derekpierre committed:
Improve error messaging for chunk size decrease factor.
Add value check for chunk size increase factor.
Add testst.

derekpierre committed:
Add test ensuring that reduced chunk size is used for scanning the remaining chunks in the block range.

derekpierre committed:
Ensure that environment variables are casted to int.

derekpierre committed:
Rename feature newsfragment from #3643 to #3648.

derekpierre committed:
Update newsfragment based on David's suggestion in original PR.

Co-authored-by: David Núñez <david@nucypher.com>

derekpierre committed:
Add acceptance test that runs using an actual alchemy polygon rpc endpoint that is under free tier.
The test only runs if the appropriate env variable is set - it is automatically set for CI.

cygnusv committed:
Log resolution of validators

cygnusv committed:
Ferveo-level test that aggregation fails if ill-formed

derekpierre committed:
Merge pull request #3647 from derekpierre/event-scanner-robustness

Event scanner robustness

derekpierre committed:
Merge pull request #3648 from derekpierre/761-debian-trixie

Use debian trixie release for base node images

cygnusv committed:
Newsfragment for #3640

cygnusv committed:
Restrict cryptography dependency to >=44.0.1

cygnusv committed:
Remove support for python 3.9 – Min version is now 3.10

cygnusv committed:
Relock python dependencies

cygnusv committed:
Newsfragment for #3649

derekpierre committed:
Manually remove pychalk, snaptime dependencies from dev-/requirements.txt files since we use our own versions `nucypher-pychalk` and `nucypher-snaptime`.

derekpierre committed:
Fix CLI unit test now that click v8.3.0 is used. When --help is provided the exit code is 0, but when not provided and help is still shown the exit code is 2.
Update test to be clearer regarding what it is doing.

derekpierre committed:
Rename Test* class to prevent warning shown during test execution. Technically test classes without tests in them should no start with "Test".

derekpierre committed:
Fix incorrect use of filepath using a test module causing tests to fail. This code worked before but is now failing with updated dependencies.

derekpierre committed:
Don't order the cohort of ursulas used for tests. Ordering is only needed when initializing a ritual since ordering is enforced by the Coordinator contract. All other uses of the cohort follows the order in the contract which can change due to handover.

derekpierre committed:
Properly clear filepath fixture.

derekpierre committed:
Fix inputs to cli runner tests.
In many cases, the operator needed to be selected from a list of operators - which seemed to be a change due to newer dependencies; before only one signer account was available, but now there are multiple.
Additionally, there were some unnecessary extra inputs e.g. 'Y' for confirmation, that were not actually needed; likely the test got somewhat stale as changes were made, but the test still did not fail since extra input does not cause a failure.

manumonti commented:
# Test report

These tests were performed with Tapir nodes running the code proposed in #3647, #3648, #3640 and #3649. Last commit [a54a2ed](https://github.com/nucypher/nucypher/pull/3649/commits/a54a2ed424ff93b6c7c8237afdecc65068ca4ec1)

## 1. Regular DKG:
### 1.1 Running a 3-of-5 ritual on `tapir` ✅
Deployed ritual 18: 
Initiate ritual tx: https://amoy.polygonscan.com//tx/0x1437734825af67067920705ee9fdee28dfb8e3254f0bb12fd3bbbd0828f6d849
End ritual tx: https://amoy.polygonscan.com/tx/0x78ebe4429d29806616a1bd30edce893fd757f48d26c59860ed2c825486a92189

Logs from tapir-1:
```
[...]
2025-09-23T20:02:55+0000 [RitualTracker#debug] (0x91d12A) Scanning events in block range 26800010 - 26800090
2025-09-23T20:02:55+0000 [EventActuator#debug] Scanning for blocks: 26800010 - 26800070, chunk size 60, last chunk scan took 0, last logs found 0
2025-09-23T20:02:55+0000 [EventActuator#debug] Querying eth_getLogs with the following parameters: {'address': '0xE690b6bCC0616Dc5294fF84ff4e00335cA52C388', 'topics': [['0xca79a3f8fdffa27f8c0a30733144db5a5bd2663f1f2561d8d665bf4da6a3dfbe', '0xca0aa5e957e38800cdd05b9f7ded91ec070e45fecf393e57a3f302f7f76e704e', '0x8a5f4f845317cfb9a694e9f1345b74d1833d84574d3858446e56a931b2c1a237', '0xa4e3b0c0b125669bbe2cfcbf57b6b13b68dc908d343dde9bdf0df75081ae403f', '0xbcca4dbb1c4bb22a406b6aa7fb93ded17778d2f20c772aa043bc11b515055278', '0x3dbc439bc92163779139421f9d1c71686b64b6e7c5ad6a368fe853dfceb99d3d']], 'fromBlock': 26800010, 'toBlock': 26800070}
2025-09-23T20:02:56+0000 [operator#debug] performing round 1 of DKG ritual #18 from blocktime 1758657728 with authority 0x3B42d26E19FF860bC4dEbB920DD8caA53F93c600.
2025-09-23T20:02:56+0000 [operator#debug] Validator #0: 0x05Be6D76d2282D24691E28E3Dc1c1A9709d70fa1, public key: a460ee662f
2025-09-23T20:02:56+0000 [operator#debug] Validator #1: 0x18F3a9ae64339E4FcfeBe1ac89Bc51aC3c83C22E, public key: 86287dc35e
2025-09-23T20:02:56+0000 [operator#debug] Validator #2: 0xA7165c0229544c84b417e53a1D3ab717EA4b4587, public key: 95de942ead
2025-09-23T20:02:56+0000 [operator#debug] Validator #3: 0xcbE2F626d84c556AbA674FABBbBDdbED6B39d87b, public key: 962a79ef82
2025-09-23T20:02:56+0000 [operator#debug] Validator #4: 0xd274f0060256c186479f2b9f51615003cbcd19E6, public key: a19df7e35c
2025-09-23T20:02:56+0000 [ferveo-dkg#debug] Initialized DKG backend for 3/5 nodes: 0x05Be, 0x18F3, 0xA716, 0xcbE2, 0xd274
2025-09-23T20:02:56+0000 [Blockchain#debug] Gas limit for this TX was increased from 638449 to 734217, using a multiplier of 1.15.
[...]
2025-09-23T20:02:57+0000 [operator#debug] 0x91d12A submitted a transcript for DKG ritual #18 (2/5) with authority 0x3B42d26E19FF860bC4dEbB920DD8caA53F93c600.
2025-09-23T20:02:57+0000 [EventActuator#debug] Processing event StartRitual, block: 26800067
Broadcasted EIP-1559 async tx 0 with TXHASH 0xedff723187b854051a093f1fcb61ebf531aea00d14f22bda97561b40a40ff8cf (0.019089642048458322 @ 26.000000066 gwei)
2025-09-23T20:02:57+0000 [stdoutemitter#info] Broadcasted EIP-1559 async tx 0 with TXHASH 0xedff723187b854051a093f1fcb61ebf531aea00d14f22bda97561b40a40ff8cf (0.019089642048458322 @ 26.000000066 gwei)
2025-09-23T20:02:57+0000 [EventActuator#debug] Scanning for blocks: 26800071 - 26800090, chunk size 60, last chunk scan took 1.3754775524139404, last logs found 0
2025-09-23T20:02:57+0000 [EventActuator#debug] Querying eth_getLogs with the following parameters: {'address': '0xE690b6bCC0616Dc5294fF84ff4e00335cA52C388', 'topics': [['0xca79a3f8fdffa27f8c0a30733144db5a5bd2663f1f2561d8d665bf4da6a3dfbe', '0xca0aa5e957e38800cdd05b9f7ded91ec070e45fecf393e57a3f302f7f76e704e', '0x8a5f4f845317cfb9a694e9f1345b74d1833d84574d3858446e56a931b2c1a237', '0xa4e3b0c0b125669bbe2cfcbf57b6b13b68dc908d343dde9bdf0df75081ae403f', '0xbcca4dbb1c4bb22a406b6aa7fb93ded17778d2f20c772aa043bc11b515055278', '0x3dbc439bc92163779139421f9d1c71686b64b6e7c5ad6a368fe853dfceb99d3d']], 'fromBlock': 26800071, 'toBlock': 26800090}
2025-09-23T20:02:57+0000 [RitualTracker#debug] Scanned total of 1 events, in 1.5649619102478027 seconds, total 2 chunk scans performed
[...]
2025-09-23T20:08:55+0000 [RitualTracker#debug] (0x91d12A) Scanning events in block range 26800190 - 26800270
2025-09-23T20:08:55+0000 [EventActuator#debug] Scanning for blocks: 26800190 - 26800250, chunk size 60, last chunk scan took 0, last logs found 0
2025-09-23T20:08:55+0000 [EventActuator#debug] Querying eth_getLogs with the following parameters: {'address': '0xE690b6bCC0616Dc5294fF84ff4e00335cA52C388', 'topics': [['0xca79a3f8fdffa27f8c0a30733144db5a5bd2663f1f2561d8d665bf4da6a3dfbe', '0xca0aa5e957e38800cdd05b9f7ded91ec070e45fecf393e57a3f302f7f76e704e', '0x8a5f4f845317cfb9a694e9f1345b74d1833d84574d3858446e56a931b2c1a237', '0xa4e3b0c0b125669bbe2cfcbf57b6b13b68dc908d343dde9bdf0df75081ae403f', '0xbcca4dbb1c4bb22a406b6aa7fb93ded17778d2f20c772aa043bc11b515055278', '0x3dbc439bc92163779139421f9d1c71686b64b6e7c5ad6a368fe853dfceb99d3d']], 'fromBlock': 26800190, 'toBlock': 26800250}
2025-09-23T20:08:55+0000 [RitualTracker#debug] Event 'EndRitual', does not require further action either: not participating in ritual, no corresponding action or previously handled; skipping
2025-09-23T20:08:56+0000 [EventActuator#debug] Processing event EndRitual, block: 26800193
2025-09-23T20:08:56+0000 [EventActuator#debug] Scanning for blocks: 26800251 - 26800270, chunk size 60, last chunk scan took 0.21678543090820312, last logs found 0
2025-09-23T20:08:56+0000 [EventActuator#debug] Querying eth_getLogs with the following parameters: {'address': '0xE690b6bCC0616Dc5294fF84ff4e00335cA52C388', 'topics': [['0xca79a3f8fdffa27f8c0a30733144db5a5bd2663f1f2561d8d665bf4da6a3dfbe', '0xca0aa5e957e38800cdd05b9f7ded91ec070e45fecf393e57a3f302f7f76e704e', '0x8a5f4f845317cfb9a694e9f1345b74d1833d84574d3858446e56a931b2c1a237', '0xa4e3b0c0b125669bbe2cfcbf57b6b13b68dc908d343dde9bdf0df75081ae403f', '0xbcca4dbb1c4bb22a406b6aa7fb93ded17778d2f20c772aa043bc11b515055278', '0x3dbc439bc92163779139421f9d1c71686b64b6e7c5ad6a368fe853dfceb99d3d']], 'fromBlock': 26800251, 'toBlock': 26800270}
2025-09-23T20:08:56+0000 [RitualTracker#debug] Scanned total of 1 events, in 0.3813331127166748 seconds, total 2 chunk scans performed
```

### 1.2 Encrypt/Decrypt test ✅

Logs from `tapir-1`
```
2025-09-23T20:28:40+0000 [twisted.python.log#info] "<IP_REDACTED>" - - [23/Sep/2025:20:28:39 +0000] "POST /node_metadata HTTP/1.1" 200 3822 "-" "python-requests/2.32.5"
2025-09-23T20:28:41+0000 [condition-eval#info] Evaluating access conditions {'version': '1.0.0', 'condition': {'chain': 80002, 'method': 'eth_getBalance', 'parameters': ['0x210eeAC07542F815ebB6FD6689637D8cA2689392', 'latest'], 'returnValueTest': {'comparator': '==', 'value': 0}, 'conditionType': 'rpc'}}
```

## 2. Regular DKG + handover
### 2.1 Running a 2-of-2 ritual on `tapir` ✅
Deployed ritual 19:
Initiate ritual tx: https://amoy.polygonscan.com//tx/0xe50c732cd20f3b36aaff26a8dabbbc06cd2c95ccf5e4e3db49bc7a50dfa7bd12
End ritual tx: https://amoy.polygonscan.com/tx/0x29d0282f032cca671ae1276ef1481a18c1f0ff3923158aa5dd0e94fa3782bda1

Nodes:
- `0x05Be6D76d2282D24691E28E3Dc1c1A9709d70fa1` | `0x91d12AB1EffBa82A4756ea029D40DE3fCD8f0255`
- `0xcbE2F626d84c556AbA674FABBbBDdbED6B39d87b` | `0xb057B982fB575509047e90cf5087c9B863a2022d`
### 2.2 Encrypt/Decrypt test ✅
Logs from tapir-1
```
2025-09-23T20:40:43+0000 [condition-eval#info] Evaluating access conditions {'version': '1.0.0', 'condition': {'chain': 80002, 'method': 'eth_getBalance', 'parameters': ['0x210eeAC07542F815ebB6FD6689637D8cA2689392', 'latest'], 'returnValueTest': {'comparator': '==', 'value': 0}, 'conditionType': 'rpc'}}
```
### 2.3 Handover of 1st original node ✅
Handover from `0x05Be6D76d2282D24691E28E3Dc1c1A9709d70fa1` to `0xd274f0060256c186479f2b9f51615003cbcd19E6`
Handover request tx: https://amoy.polygonscan.com//tx/0xd582692f86406154d25cd434c2343da5afd84c656f1bce2e242576a8507913b1
Post handover transcript tx: https://amoy.polygonscan.com/tx/0xbcd5dc4d1a1664cfdbad80c778b0e6655365702b8a92517a26e2f3671c9e4aa2
Post blinded share tx: https://amoy.polygonscan.com/tx/0x9ea22ab04c8e56b0ea14e937e5167feb127becf89435b5aff7de46b26f3d84b8
Finalize handover tx: https://amoy.polygonscan.com/tx/0xf8fab1610601d9d4491fe752831e5489fa56a6341d0ddd23965119b6d2d13929

Logs from departing node `0x05Be6D76d2282D24691E28E3Dc1c1A9709d70fa1`:
```
2025-09-23T20:52:55+0000 [RitualTracker#debug] Event 'HandoverRequest', does not require further action either: not participating in ritual, no corresponding action or previously handled; skipping
[...]
2025-09-23T20:52:56+0000 [operator#info] Handover applicable to node: status=2, ritual=#19, departing_validator=0x05Be6D76d2282D24691E28E3Dc1c1A9709d70fa1.
2025-09-23T20:52:56+0000 [operator#info] Handover applicable to node: status=2, ritual=#19, departing_validator=0x05Be6D76d2282D24691E28E3Dc1c1A9709d70fa1.
2025-09-23T20:52:56+0000 [operator#info] 0x91d12A producing a new blinded share for handover at ritual #19.
2025-09-23T20:52:56+0000 [operator#debug] Validator #0: 0x05Be6D76d2282D24691E28E3Dc1c1A9709d70fa1, public key: a460ee662f
2025-09-23T20:52:56+0000 [operator#debug] Validator #1: 0xcbE2F626d84c556AbA674FABBbBDdbED6B39d87b, public key: 962a79ef82
2025-09-23T20:52:56+0000 [operator#info] Resulting blinded share: 8205d4672de6836ed9f8bf8de18afd84ed6c959932fda7f808fe6fa3977e96a76d63cf37729e891f832ad406ece67215174f364731d855aa8d60dc0a9a57f917ac219a16d7e36b50b626490342fb3cda051b8c17f0e73d36321e3009014f814f (share_index: 0, start: 128)
2025-09-23T20:52:56+0000 [operator#info] Publishing handover blinded share (8205d4672de6836ed9f8) for ritual #19
[...]
2025-09-23T20:52:57+0000 [EventActuator#debug] Processing event HandoverTranscriptPosted, block: 26801588
[...]
2025-09-23T20:54:56+0000 [operator#info] Handover status (3) for ritual #19 is not in the expected state (2).
2025-09-23T20:54:56+0000 [operator#info] No action required for handover blinded share for ritual #19
[...]
2025-09-23T21:04:56+0000 [RitualTracker#debug] Event 'HandoverFinalized', does not require further action either: not participating in ritual, no corresponding action or previously handled; skipping
2025-09-23T21:04:56+0000 [EventActuator#debug] Processing event HandoverFinalized, block: 26801945
```

Logs from incoming node `0xd274f0060256c186479f2b9f51615003cbcd19E6`:
```
2025-09-23T20:52:46+0000 [operator#info] Handover applicable to node: status=1, ritual=#19, departing_validator=0x05Be6D76d2282D24691E28E3Dc1c1A9709d70fa1.
2025-09-23T20:52:46+0000 [operator#debug] Validator #0: 0x05Be6D76d2282D24691E28E3Dc1c1A9709d70fa1, public key: a460ee662f
2025-09-23T20:52:46+0000 [operator#debug] Validator #1: 0xcbE2F626d84c556AbA674FABBbBDdbED6B39d87b, public key: 962a79ef82
2025-09-23T20:52:46+0000 [operator#info] Producing handover transcript for ritual #19, departing validator 0x05Be6D76d2282D24691E28E3Dc1c1A9709d70fa1, and handover slot index 0.
2025-09-23T20:52:46+0000 [ferveo-dkg#debug] Initialized DKG backend for 2/2 nodes: 0x05Be, 0xcbE2
2025-09-23T20:52:46+0000 [operator#info] Resulting handover transcript (first 10 bytes): 00000000600000000000
2025-09-23T20:52:46+0000 [operator#info] Publishing handover transcript (00000000600000000000) for ritual #19 and departing validator 0x05Be6D76d2282D24691E28E3Dc1c1A9709d70fa1
[...]
2025-09-23T20:54:46+0000 [operator#info] Handover status (3) for ritual #19 is not in the expected state (1).
2025-09-23T20:54:46+0000 [operator#info] No action required for handover transcript for ritual #19
[...]
2025-09-23T21:06:46+0000 [RitualTracker#debug] Event 'HandoverFinalized', does not require further action either: not participating in ritual, no corresponding action or previously handled; skipping
2025-09-23T21:06:46+0000 [EventActuator#debug] Processing event HandoverFinalized, block: 26801945
```

Call to getParticipants(19) on coordinator:
```
[[0xd274f0060256c186479f2b9f51615003cbcd19E6,true,0x,0x5453536b00020000c42048ea67e080e6798ce5667b5919fffe401a2d66f04da3ba303518d6616c898731]  
[0xcbE2F626d84c556AbA674FABBbBDdbED6B39d87b,true,0x6800000000000000020000000000000096db83ceea9cfc424139b200ee662da585969f8ffc25a690854f08ce321653c7e8d8429be7292c3025c4fcbf8a2bd16887b2abb5218f0befe474e6f473a157a87b9aa29875f5ccd67f5ac40da9f0559fb7c27c49243a3163daa3309a6850074ac8000000000000000200000000000000b33cbcd11c48a82c0e97293dcf7dcf0e25d8dd4a712ef849b865ea9a8d8c1e51dadf107995c87e5871dddd61469f03a409b1c25a884202d3b468b77f7704e1612df0e73af2726427102e8dafd5cade7a20e5e5ad4843504042c0b628a08c41fd803cea20923a1d5cdc3382d2b2e7a0783498606bf53fd62f4f8dc4fd80bb74d20c32f68e628b38c292fbd7aa687b9919045f03fad299cceff3a80c173551c80900a521e141de9e50c33aa4843bf2cc9dd717da0cbe24daeb6b6e197b3c933dbb6000000000000000aff76c419b20e1d0916016b58b7047a3eae998e240de950ad1c3d47347fe74c4e83bc56b9925127a1b4f6300d7925ca510f507c4e0a408d5dcb605c63c8d4f48bf72d5c0f0a27b0591d613773c4612c866d183f4e16343124b23cd138088f044,0x5453536b00020000c42020cce78f8914f23f589eb663ca566953aaee43386310496a905913a737cfca5c]]
```
### 2.4 Encryption/Decryption test after 1st handover ✅
Logs from incoming node:
```
2025-09-23T21:10:48+0000 [condition-eval#info] Evaluating access conditions {'version': '1.0.0', 'condition': {'chain': 80002, 'method': 'eth_getBalance', 'parameters': ['0x210eeAC07542F815ebB6FD6689637D8cA2689392', 'latest'], 'returnValueTest': {'comparator': '==', 'value': 0}, 'conditionType': 'rpc'}}
```

### 2.5 Handover of the 2nd original node ✅
Handover from `0xcbE2F626d84c556AbA674FABBbBDdbED6B39d87b` to `0x18F3a9ae64339E4FcfeBe1ac89Bc51aC3c83C22E`.
Handover request tx: https://amoy.polygonscan.com//tx/0x78077acc800221b7dd1641ea49a009a72dad8226d34d5314d046b8e31111dbc8
Post Handover transcript tx: https://amoy.polygonscan.com/tx/0x8d9671882055f60b5c0efbdd27241dfcc3ecd71ecf78069580bac7e2960d258d
Post blinded share tx: https://amoy.polygonscan.com/tx/0xf18bef9dd6ef6a1774d14a8ab4f1b4e90bbc787b60c5bd7bea2fcfea93b68acf
Finalize handover tx: https://amoy.polygonscan.com//tx/0x7b88147bad94851245064ebf110064f18b324d00ce9dbd17271997d06e4b8b49


Logs from departing node `0xcbE2F626d84c556AbA674FABBbBDdbED6B39d87b`:
```
2025-09-23T21:19:31+0000 [RitualTracker#debug] Event 'HandoverRequest', does not require further action either: not participating in ritual, no corresponding action or previously handled; skipping
2025-09-23T21:19:31+0000 [EventActuator#debug] Processing event HandoverRequest, block: 26802375
2025-09-23T21:19:31+0000 [operator#info] Handover applicable to node: status=2, ritual=#19, departing_validator=0xcbE2F626d84c556AbA674FABBbBDdbED6B39d87b.
2025-09-23T21:19:31+0000 [operator#info] Handover applicable to node: status=2, ritual=#19, departing_validator=0xcbE2F626d84c556AbA674FABBbBDdbED6B39d87b.
2025-09-23T21:19:31+0000 [operator#info] 0xb057B9 producing a new blinded share for handover at ritual #19.
2025-09-23T21:19:31+0000 [operator#debug] Validator #0: 0xd274f0060256c186479f2b9f51615003cbcd19E6, public key: a19df7e35c
2025-09-23T21:19:31+0000 [operator#debug] Validator #1: 0xcbE2F626d84c556AbA674FABBbBDdbED6B39d87b, public key: 962a79ef82
2025-09-23T21:19:31+0000 [operator#info] Resulting blinded share: 8543d4cc082e83f8b0f410eea16731b59d230766a77bcaad44854481f6890581174e7c231b1a1e784ced9e07bc837aaa18381209d9984c04c6a0276ef16240eba9ec6b0a6d2d49d8b65654fa2a8b71262bf8021f6f7eec963d12ea1ecec346a7 (share_index: 1, start: 224)
2025-09-23T21:19:31+0000 [operator#info] Publishing handover blinded share (8543d4cc082e83f8b0f4) for ritual #19
[...]
2025-09-23T21:19:32+0000 [operator#info] 0xb057B9 created a handover blinded share for DKG ritual #19.
2025-09-23T21:19:32+0000 [EventActuator#debug] Processing event HandoverTranscriptPosted, block: 26802385
[...]
2025-09-23T21:21:31+0000 [RitualTracker#debug] Event 'HandoverRequest', does not require further action either: not participating in ritual, no corresponding action or previously handled; skipping
2025-09-23T21:21:31+0000 [EventActuator#debug] Processing event HandoverRequest, block: 26802375
2025-09-23T21:21:31+0000 [operator#info] Handover status (3) for ritual #19 is not in the expected state (2).
2025-09-23T21:21:31+0000 [operator#info] No action required for handover blinded share for ritual #19
2025-09-23T21:21:31+0000 [EventActuator#debug] Processing event HandoverTranscriptPosted, block: 26802385
[...]
2025-09-23T21:23:31+0000 [RitualTracker#debug] Event 'HandoverFinalized', does not require further action either: not participating in ritual, no corresponding action or previously handled; skipping
2025-09-23T21:23:31+0000 [EventActuator#debug] Processing event HandoverFinalized, block: 26802503
```

Logs from incoming node `0x18F3a9ae64339E4FcfeBe1ac89Bc51aC3c83C22E`:
```
2025-09-23T21:19:20+0000 [operator#info] Handover applicable to node: status=1, ritual=#19, departing_validator=0xcbE2F626d84c556AbA674FABBbBDdbED6B39d87b.
2025-09-23T21:19:20+0000 [operator#debug] Validator #0: 0xd274f0060256c186479f2b9f51615003cbcd19E6, public key: a19df7e35c
2025-09-23T21:19:20+0000 [operator#debug] Validator #1: 0xcbE2F626d84c556AbA674FABBbBDdbED6B39d87b, public key: 962a79ef82
2025-09-23T21:19:20+0000 [operator#info] Producing handover transcript for ritual #19, departing validator 0xcbE2F626d84c556AbA674FABBbBDdbED6B39d87b, and handover slot index 1.
2025-09-23T21:19:20+0000 [ferveo-dkg#debug] Initialized DKG backend for 2/2 nodes: 0xd274, 0xcbE2
2025-09-23T21:19:20+0000 [operator#info] Resulting handover transcript (first 10 bytes): 01000000600000000000
2025-09-23T21:19:20+0000 [operator#info] Publishing handover transcript (01000000600000000000) for ritual #19 and departing validator 0xcbE2F626d84c556AbA674FABBbBDdbED6B39d87b
[...]
2025-09-23T21:19:20+0000 [operator#info] 0x131617 created a handover transcript for DKG ritual #19 and departing validator 0xcbE2F626d84c556AbA674FABBbBDdbED6B39d87b.
2025-09-23T21:19:20+0000 [EventActuator#debug] Processing event HandoverRequest, block: 26802375
[...]
2025-09-23T21:21:19+0000 [operator#info] Handover status (3) for ritual #19 is not in the expected state (1).
2025-09-23T21:21:19+0000 [operator#info] No action required for handover transcript for ritual #19
2025-09-23T21:21:19+0000 [EventActuator#debug] Processing event HandoverRequest, block: 26802375
2025-09-23T21:21:19+0000 [RitualTracker#debug] Event 'HandoverTranscriptPosted', does not require further action either: not participating in ritual, no corresponding action or previously handled; skipping
2025-09-23T21:21:19+0000 [EventActuator#debug] Processing event HandoverTranscriptPosted, block: 26802385
[...]
2025-09-23T21:25:19+0000 [RitualTracker#debug] Event 'HandoverFinalized', does not require further action either: not participating in ritual, no corresponding action or previously handled; skipping
2025-09-23T21:25:19+0000 [EventActuator#debug] Processing event HandoverFinalized, block: 26802503
```

Call to getParticipants(19) on coordinator:
```
[[0xd274f0060256c186479f2b9f51615003cbcd19E6,true,0x,0x5453536b00020000c42048ea67e080e6798ce5667b5919fffe401a2d66f04da3ba303518d6616c898731]  
[0x18F3a9ae64339E4FcfeBe1ac89Bc51aC3c83C22E,true,0x,0x5453536b00020000c420fb5d2860a969d913685ee27a0e956536eef3ba8d15ca349f36a0becc93fd5f20]]
```

### 2.5 Encryption/Decryption test after 2nd handover ✅
Logs from incoming node 
```
2025-09-23T21:29:26+0000 [condition-eval#info] Evaluating access conditions {'version': '1.0.0', 'condition': {'chain': 80002, 'method': 'eth_getBalance', 'parameters': ['0x210eeAC07542F815ebB6FD6689637D8cA2689392', 'latest'], 'returnValueTest': {'comparator': '==', 'value': 0}, 'conditionType': 'rpc'}}
```

@derekpierre @cygnusv

derekpierre committed:
Clean up cli help test to simplify exit code check and sub-command entry point printing in help output.

derekpierre committed:
Merge pull request #3640 from cygnusv/post-handover

Post handover: logging improvements

derekpierre committed:
Merge pull request #3649 from cygnusv/relock

Drop support for Python 3.9 and Dependency Relock

manumonti reviewed (approved):
LGTM! let's do it

derekpierre committed:
Fix very unlikely but transient bug where dkg storage contains stored ritual but not stored validators. These are done in separate steps by nodes, with the ritual stored first and then validators. Therefore, there is a very small race condition window where the ritual is set but the validators are not.
To resolve this, we reverse the order in the test i.e. check validator first to determine participation and then check ritual.

derekpierre committed:
Add dev newsfragment for #3650.

derekpierre committed:
Merge pull request #3650 from derekpierre/fix-transient-failure

Fix transient test failure
